gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt

```diff
@@ -279,14 +292,42 @@ internal open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
     }
   }
 
-  private fun disableIncrementalKotlinCompilation(variant: Variant) {
+  private fun disableIncrementalKotlinCompilation(
+    variant: Variant,
+    compileTaskProvider: TaskProvider<out KotlinCompilationTask<*>>,
+    irMergesFile: Provider<File>,
+  ) {
+
+    fun mergingIsDisabled(): Boolean {
+      return variant.variantFilter.generateDaggerFactoriesOnly ||
+        variant.variantFilter.disableComponentMerging
+    }
+
+    compileTaskProvider.configure { task ->
+
+      if (variant.variantFilter.trackSourceFiles && !mergingIsDisabled()) {
+        if (task is AbstractKotlinCompile<*>) {
+          // The `ir-merges.txt` file indicates that the last compilation resulted in adding module
+          // arguments to an annotation or adding interface supertypes.  Those changes do not affect
+          // the output .class files, which means that Kotlin's incremental logic doesn't properly
+          // track the changes.  So, we disable incremental compilation whenever this file exists.
+          //
+          // Note that this is being done at configuration time,
+          // and not inside a `doFirst { }` block.  This is because the `doFirst { }` block would
+          // execute after build cache restoration has taken place.  We only need to disable
+          // incremental logic if the previous compilation happened on this machine and the
+          // resultant ModuleDescriptors are already in memory.  If the files need to be restored
+          // from cache, then the incremental logic we're avoiding won't happen anyway.
+          task.incremental = task.incremental && !irMergesFile.get().exists()
+        }
+      }
+    }
+
     variant.project.pluginManager.withPlugin(KAPT_PLUGIN_ID) {
       variant.project
         .namedLazy<KaptGenerateStubsTask>(variant.stubsTaskName) { stubsTaskProvider ->
-          if (!variant.variantFilter.generateDaggerFactoriesOnly &&
-            !variant.variantFilter.disableComponentMerging
-          ) {
-            stubsTaskProvider.configure { stubsTask ->
+          stubsTaskProvider.configure { stubsTask ->
+            if (!mergingIsDisabled()) {
```

Comment:
Is there any downside (or benefit) to moving this check into the config block instead of avoiding the config block entirely?

Reply:
When the check happens outside the config block, it reads the variant's settings eagerly, as soon as the compilation is configured.  When the check happens inside the stub gen task's config block, it has a better chance of catching any late changes made by a convention plugin or something.